<a href="https://www.kaggle.com/code/veeraj16/eeg-classification?scriptVersionId=328004005" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
# Reproducible, leak-free BCI Competition IV 2a pipeline.
# Installs are quiet unless something actually fails.
import os, sys, subprocess, random, warnings, contextlib, io
warnings.filterwarnings('ignore')
os.environ['PYTHONHASHSEED'] = '2026'
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
os.environ['TF_DETERMINISTIC_OPS'] = '1'
random.seed(2026)

packages = [
    'moabb==1.1.1', 'pyriemann==0.7', 'pymatreader', 'docstring-inheritance',
    'pooch', 'memory-profiler', 'edflib-python', 'edfio', 'mne-bids', 'pybv',
]
install = subprocess.run(
    [sys.executable, '-m', 'pip', 'install', '-q', '--no-deps', *packages],
    stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True,
)
if install.returncode != 0:
    print(install.stdout[-3000:])
    raise RuntimeError('Dependency installation failed')

import numpy as np
import pandas as pd
import scipy, sklearn, mne, moabb
mne.set_log_level('ERROR')

from sklearn.preprocessing import LabelEncoder
from moabb.datasets import BNCI2014_001
from moabb.paradigms import MotorImagery

with contextlib.redirect_stdout(io.StringIO()), contextlib.redirect_stderr(io.StringIO()):
    dataset = BNCI2014_001()
    paradigm = MotorImagery(n_classes=4, channels=None, resample=250, fmin=4, fmax=40, tmin=0.0, tmax=3.5)
    X, labels, meta = paradigm.get_data(dataset=dataset, subjects=list(range(1, 10)))

X = np.asarray(X, dtype=np.float64)
subjects = meta['subject'].to_numpy()
sessions = meta['session'].to_numpy()
le = LabelEncoder()
y = le.fit_transform(labels)
assert X.shape == (5184, 22, 876), X.shape
print(f'Data ready: X={X.shape}, subjects={len(pd.unique(subjects))}, classes={list(le.classes_)}')

SHA256 hash of downloaded file: 054f02e70cf9c4ada1517e9b9864f45407939c1062c6793516585c6f511d0325
Use this value as the 'known_hash' argument of 'pooch.retrieve' to ensure that the file hasn't changed if it is downloaded again in the future.
SHA256 hash of downloaded file: 53d415f39c3d7b0c88b894d7b08d99bcdfe855ede63831d3691af1a45607fb62
Use this value as the 'known_hash' argument of 'pooch.retrieve' to ensure that the file hasn't changed if it is downloaded again in the future.
SHA256 hash of downloaded file: 5ddd5cb520b1692c3ba1363f48d98f58f0e46f3699ee50d749947950fc39db27
Use this value as the 'known_hash' argument of 'pooch.retrieve' to ensure that the file hasn't changed if it is downloaded again in the future.
SHA256 hash of downloaded file: d63c454005d3a9b41d8440629482e855afc823339bdd0b5721842a7ee9cc7b12
Use this value as the 'known_hash' argument of 'pooch.retrieve' to ensure that the file hasn't changed if it is downloaded again in the future.
SHA256 hash of downloaded file: 7e73

Data ready: X=(5184, 22, 876), subjects=9, classes=['feet', 'left_hand', 'right_hand', 'tongue']


In [2]:
# Honest cross-subject baseline: held-out subjects, no windows before split
from sklearn.model_selection import GroupShuffleSplit
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score, cohen_kappa_score, confusion_matrix
from pyriemann.estimation import Covariances
from pyriemann.tangentspace import TangentSpace

# 6 train subjects, 1 validation subject, 2 test subjects.
gss_test = GroupShuffleSplit(n_splits=1, test_size=2/9, random_state=42)
tv_idx, cs_test_idx = next(gss_test.split(X, y, groups=subjects))
gss_val = GroupShuffleSplit(n_splits=1, test_size=1/7, random_state=42)
tr_rel, va_rel = next(gss_val.split(X[tv_idx], y[tv_idx], groups=subjects[tv_idx]))
cs_train_idx, cs_val_idx = tv_idx[tr_rel], tv_idx[va_rel]

assert set(subjects[cs_train_idx]).isdisjoint(subjects[cs_val_idx])
assert set(subjects[cs_train_idx]).isdisjoint(subjects[cs_test_idx])
assert set(subjects[cs_val_idx]).isdisjoint(subjects[cs_test_idx])

def train_channel_standardize(train_idx, *idx_sets):
    mu = X[train_idx].mean(axis=(0, 2), keepdims=True)
    sd = X[train_idx].std(axis=(0, 2), keepdims=True) + 1e-6
    return [((X[idx] - mu) / sd).astype('float64') for idx in idx_sets]

Xcs_tr, Xcs_va, Xcs_te = train_channel_standardize(cs_train_idx, cs_train_idx, cs_val_idx, cs_test_idx)
cs_model = make_pipeline(
    Covariances(estimator='oas'),
    TangentSpace(metric='riemann'),
    StandardScaler(),
    LogisticRegression(max_iter=5000, class_weight='balanced', C=0.1),
)
cs_model.fit(Xcs_tr, y[cs_train_idx])
cs_pred = cs_model.predict(Xcs_te)

cross_subject_metrics = {
    'protocol': 'held-out subjects',
    'train_subjects': sorted(pd.unique(subjects[cs_train_idx]).tolist()),
    'val_subjects': sorted(pd.unique(subjects[cs_val_idx]).tolist()),
    'test_subjects': sorted(pd.unique(subjects[cs_test_idx]).tolist()),
    'accuracy': accuracy_score(y[cs_test_idx], cs_pred),
    'macro_f1': f1_score(y[cs_test_idx], cs_pred, average='macro'),
    'kappa': cohen_kappa_score(y[cs_test_idx], cs_pred),
}

display(pd.DataFrame([{k: v for k, v in cross_subject_metrics.items() if not k.endswith('subjects')}]).round(4))
print('Subjects:', {k: cross_subject_metrics[k] for k in ['train_subjects', 'val_subjects', 'test_subjects']})

,protocol,accuracy,macro_f1,kappa
0,held-out subjects,0.316,0.2621,0.088


Subjects: {'train_subjects': [3, 4, 5, 6, 7, 9], 'val_subjects': [1], 'test_subjects': [2, 8]}


In [3]:
# Subject-dependent session split: FBCSP + logistic regression
import contextlib, io
from scipy.signal import butter, sosfiltfilt
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score, cohen_kappa_score, confusion_matrix
from mne.decoding import CSP

bands = [(4,8), (8,12), (12,16), (16,20), (20,24), (24,30), (30,40)]
Cs = [0.03, 0.1, 0.3, 1.0, 3.0]

def bandpass(x, low, high, fs=250.0, order=4):
    sos = butter(order, [low, high], btype='bandpass', fs=fs, output='sos')
    return sosfiltfilt(sos, x, axis=-1).astype('float64')

def fbcsp_features(Xtr, ytr, Xva, Xte, n_components=4):
    tr, va, te = [], [], []
    for low, high in bands:
        with contextlib.redirect_stdout(io.StringIO()), contextlib.redirect_stderr(io.StringIO()):
            csp = CSP(n_components=n_components, reg='ledoit_wolf', log=True, norm_trace=False)
            tr.append(csp.fit_transform(bandpass(Xtr, low, high), ytr))
            va.append(csp.transform(bandpass(Xva, low, high)))
            te.append(csp.transform(bandpass(Xte, low, high)))
    return np.concatenate(tr, axis=1), np.concatenate(va, axis=1), np.concatenate(te, axis=1)

rows, y_true_all, y_pred_all = [], [], []
for subj in sorted(pd.unique(subjects)):
    train_pool = np.where((subjects == subj) & (sessions == '0train'))[0]
    test_idx = np.where((subjects == subj) & (sessions == '1test'))[0]
    split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=100 + int(subj))
    tr_rel, va_rel = next(split.split(X[train_pool], y[train_pool]))
    tr_idx, va_idx = train_pool[tr_rel], train_pool[va_rel]

    Ftr, Fva, Fte = fbcsp_features(X[tr_idx], y[tr_idx], X[va_idx], X[test_idx])
    best = None
    for C in Cs:
        clf = make_pipeline(StandardScaler(), LogisticRegression(C=C, max_iter=5000, class_weight='balanced'))
        clf.fit(Ftr, y[tr_idx])
        val_acc = accuracy_score(y[va_idx], clf.predict(Fva))
        if best is None or val_acc > best[0]:
            best = (val_acc, C, clf)

    pred = best[2].predict(Fte)
    y_true_all.append(y[test_idx]); y_pred_all.append(pred)
    rows.append({'subject': int(subj), 'val_acc': best[0], 'C': best[1], 'test_acc': accuracy_score(y[test_idx], pred)})

y_true_fbcsp = np.concatenate(y_true_all)
y_pred_fbcsp = np.concatenate(y_pred_all)
fbcsp_metrics = {
    'protocol': 'subject session split',
    'model': 'FBCSP + LR',
    'accuracy': accuracy_score(y_true_fbcsp, y_pred_fbcsp),
    'macro_f1': f1_score(y_true_fbcsp, y_pred_fbcsp, average='macro'),
    'kappa': cohen_kappa_score(y_true_fbcsp, y_pred_fbcsp),
}

display(pd.DataFrame([fbcsp_metrics]).round(4))

,protocol,model,accuracy,macro_f1,kappa
0,subject session split,FBCSP + LR,0.5949,0.5948,0.4599


In [4]:
# Neural baseline: compact EEGNet trained in a captured subprocess
import json, subprocess, sys, textwrap
from pathlib import Path

payload_path = Path('/kaggle/working/eegnet_payload.npz')
np.savez(payload_path, X=X.astype('float32'), y=y, subjects=subjects, sessions=sessions)

child_code = r'''
import os, random, json, warnings
warnings.filterwarnings('ignore')
os.environ['PYTHONHASHSEED'] = '2026'
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
os.environ['TF_DETERMINISTIC_OPS'] = '1'
random.seed(2026)

import numpy as np
import tensorflow as tf
from tensorflow.keras import layers, models, callbacks, regularizers
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.metrics import accuracy_score, f1_score, cohen_kappa_score

np.random.seed(2026); tf.random.set_seed(2026)
tf.keras.backend.clear_session()
data = np.load('/kaggle/working/eegnet_payload.npz', allow_pickle=True)
X, y, subjects, sessions = data['X'], data['y'], data['subjects'], data['sessions']

train_pool = np.where(sessions == '0train')[0]
test_idx = np.where(sessions == '1test')[0]
strata = np.array([f'S{s}_Y{yy}' for s, yy in zip(subjects[train_pool], y[train_pool])])
tr_rel, va_rel = next(StratifiedShuffleSplit(n_splits=1, test_size=0.15, random_state=2026).split(X[train_pool], strata))
tr_idx, va_idx = train_pool[tr_rel], train_pool[va_rel]

mu = X[tr_idx].mean(axis=(0, 2), keepdims=True)
sigma = X[tr_idx].std(axis=(0, 2), keepdims=True) + 1e-6
norm = lambda z: ((z - mu) / sigma).astype('float32')[..., None]
Xtr, Xva, Xte = norm(X[tr_idx]), norm(X[va_idx]), norm(X[test_idx])
ytr = tf.keras.utils.to_categorical(y[tr_idx], 4)
yva = tf.keras.utils.to_categorical(y[va_idx], 4)

def build_eegnet(n_ch=22, n_t=876, n_classes=4):
    inp = layers.Input(shape=(n_ch, n_t, 1))
    x = layers.Conv2D(8, (1, 64), padding='same', use_bias=False)(inp)
    x = layers.BatchNormalization()(x)
    x = layers.DepthwiseConv2D((n_ch, 1), depth_multiplier=2, use_bias=False)(x)
    x = layers.BatchNormalization()(x); x = layers.Activation('elu')(x)
    x = layers.AveragePooling2D((1, 4))(x); x = layers.Dropout(0.35)(x)
    x = layers.SeparableConv2D(16, (1, 16), padding='same', use_bias=False)(x)
    x = layers.BatchNormalization()(x); x = layers.Activation('elu')(x)
    x = layers.AveragePooling2D((1, 8))(x); x = layers.Dropout(0.35)(x)
    x = layers.Flatten()(x)
    x = layers.Dense(64, activation='elu', kernel_regularizer=regularizers.l2(1e-4))(x)
    x = layers.Dropout(0.4)(x)
    return models.Model(inp, layers.Dense(n_classes, activation='softmax')(x))

model = build_eegnet()
model.compile(optimizer=tf.keras.optimizers.Adam(8e-4), loss='categorical_crossentropy', metrics=['accuracy'])
history = model.fit(
    Xtr, ytr, validation_data=(Xva, yva), epochs=80, batch_size=64, verbose=0,
    callbacks=[
        callbacks.EarlyStopping(monitor='val_accuracy', mode='max', patience=14, restore_best_weights=True),
        callbacks.ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=6, min_lr=1e-5),
    ],
)
pred = model.predict(Xte, batch_size=128, verbose=0).argmax(axis=1)
metrics = {
    'protocol': 'pooled session split',
    'model': 'Compact EEGNet',
    'accuracy': float(accuracy_score(y[test_idx], pred)),
    'macro_f1': float(f1_score(y[test_idx], pred, average='macro')),
    'kappa': float(cohen_kappa_score(y[test_idx], pred)),
    'epochs': int(len(history.history['loss'])),
    'best_val_acc': float(max(history.history['val_accuracy'])),
}
print(json.dumps(metrics))
'''

proc = subprocess.run([sys.executable, '-c', child_code], capture_output=True, text=True, timeout=600)
if proc.returncode != 0:
    print(proc.stdout[-1000:])
    print(proc.stderr[-3000:])
    raise RuntimeError('EEGNet subprocess failed')

eegnet_metrics = json.loads(proc.stdout.strip().splitlines()[-1])
summary = {
    'trials': X.shape[0], 'channels': X.shape[1], 'samples_per_trial': X.shape[2],
    'subjects': len(pd.unique(subjects)), 'sessions': ', '.join(map(str, sorted(pd.unique(sessions)))),
    'classes': ', '.join(le.classes_),
}
results = pd.DataFrame([
    {'protocol': cross_subject_metrics['protocol'], 'model': 'Riemannian TS-LR',
     'accuracy': cross_subject_metrics['accuracy'], 'macro_f1': cross_subject_metrics['macro_f1'], 'kappa': cross_subject_metrics['kappa']},
    fbcsp_metrics,
    eegnet_metrics,
])

display(pd.DataFrame([eegnet_metrics]).round(4))

,protocol,model,accuracy,macro_f1,kappa,epochs,best_val_acc
0,pooled session split,Compact EEGNet,0.5548,0.5515,0.4064,80,0.5913


## Reviewer Notes

This version uses the official BNCI2014_001 / BCI Competition IV 2a data through MOABB, not the cropped Kaggle CSV. It fixes the main review blockers: no epoch-only grouping, no pre-split windows, no test-as-validation, train-only per-channel normalization, subject/session-aware splits, and no physically invalid FFT slicing.

| Protocol | Model | Accuracy | Macro F1 | Cohen kappa |
|---|---:|---:|---:|---:|
| Held-out subjects | Riemannian TS-LR | 0.3160 | 0.2621 | 0.0880 |
| Subject session split | FBCSP + LR | 0.5949 | 0.5948 | 0.4599 |
| Pooled session split | Compact EEGNet | 0.5548 | 0.5515 | 0.4064 |

Held-out-subject and subject/session results are different tasks, so they should be reported separately.